File: January Challenge/Python/Minimum falling path sum.py
class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        n, m = len(matrix), len(matrix[0])
        dp = [[int(1e9) for _ in range(m+1)] for _ in range(n+1)]

        def func(r, c, dp):
            for j in range(m):
                dp[n-1][j] = matrix[n-1][j]

            for i in range(n-2, -1, -1):
                for j in range(m-1, -1, -1): 
                    ld, rd, below = int(1e9), int(1e9), int(1e9)
                    if i + 1 < n and j - 1 >= 0:  
                        ld = matrix[i][j] + dp[i+1][j-1]

                    if i + 1 < n and j + 1 < m:  
                        rd = matrix[i][j] + dp[i+1][j+1]

                    if i + 1 < n: 
                        below = matrix[i][j] + dp[i+1][j]

                    dp[i][j] = min(ld, rd, below)
            return dp[r][c]

        mini = int(1e9)
        for i in range(m):
            res = func(0, i, dp)
            mini = min(mini, res)
        return mini

File: January Challenge/Java/Minimum falling path sum.java
import java.util.Arrays;

class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int n = matrix.length;
        int m = matrix[0].length;
        int mini = Integer.MAX_VALUE;
        
        for (int i = 0; i < m; i++) {
            int res = func(matrix, 0, i, n, m);
            mini = Math.min(mini, res);
        }
        return mini;
    }

    public int func(int[][] matrix, int r, int c, int n, int m) {
        int[] prev = new int[m + 1];
        Arrays.fill(prev, Integer.MAX_VALUE);

        for (int j = 0; j < m; j++) {
            prev[j] = matrix[n - 1][j];
        }

        for (int i = n - 2; i >= 0; i--) {
            int[] curr = new int[m + 1];
            Arrays.fill(curr, Integer.MAX_VALUE);

            for (int j = m - 1; j >= 0; j--) {
                int ld = Integer.MAX_VALUE, rd = Integer.MAX_VALUE, below = Integer.MAX_VALUE;
                
                if (i + 1 < n && j - 1 >= 0) {
                    ld = matrix[i][j] + prev[j - 1];
                }
                if (i + 1 < n && j + 1 < m) {
                    rd = matrix[i][j] + prev[j + 1];
                }
                if (i + 1 < n) {
                    below = matrix[i][j] + prev[j];
                }

                curr[j] = Math.min(Math.min(ld, rd), below);
            }

            System.arraycopy(curr, 0, prev, 0, curr.length);
        }
        return prev[c];
    }
}
